### PR TITLE
UAR-91: OE number validation

### PR DESCRIPTION
**Pull request by jkayyala**
### JIRA link

 [UAR-91](https://companieshouse.atlassian.net/browse/UAR-91)

### Change description

Added validation and redirect to UAR-91

### Work checklist

- [x] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/validation/error.messages.ts
Comment (AndreKramer7654) [inline src/validation/error.messages.ts:110]: MAX is not really fitting here INVALID_OE_LENGTH?

File: src/validation/overseas.entity.query.validation.ts
Comment (AndreKramer7654) [inline src/validation/overseas.entity.query.validation.ts:6]: Why two withMessage?

**Comment by corybudd-ch** [comment]
I think this branch now needs re-based .. and hopefully repoints at main. Maybe it won't because it needed rebased first .. not sure lol. Either way needs rebased and pointed at main then it'll be easier to review

File: src/validation/overseas.entity.query.validation.ts
Comment (AndreKramer7654) [inline src/validation/overseas.entity.query.validation.ts:6]: isLength(8) may work for us

**Review by AndreKramer7654 — APPROVED** [review]

File: src/middleware/authentication.middleware.ts
Comment (AndreKramer7654) [inline src/middleware/authentication.middleware.ts:15]: Should we test for UPDATE_LANDING_URL and turn this if/then arround so that the default is registration?

**Review by AndreKramer7654 — APPROVED** [review]

**Review by corybudd-ch — APPROVED** [review]

File: src/validation/error.messages.ts
Comment (mwestacott) [inline src/validation/error.messages.ts:20]: Constant name looks wrong as message mentions 'number' not 'name'.

File: src/validation/error.messages.ts
Comment (mwestacott) [inline src/validation/error.messages.ts:137]: Would suggest getting 'number' into the constant name (e.g. INVALID_OE_NUMBER_LENGTH) as an overseas entity doesn't really have a length.

Also, wondering if message is correct/accurate enough? Does this 'number' only contain numbers and no letters/special chars? If yes (just numbers), could it maybe say "OE number must be 8 digits"?

File: src/validation/overseas.entity.query.validation.ts
Comment (mwestacott) [inline src/validation/overseas.entity.query.validation.ts:6]: You might also need to check for invalid characters in this field - see other validation examples.

File: test/controllers/update/overseas.entity.query.controller.spec.ts
Comment (mwestacott) [inline test/controllers/update/overseas.entity.query.controller.spec.ts:107]: I would add a `.not.toContain(ErrorMessages.OE_QUERY_NAME)` here to specifically only check for the one message. Example of this elsewhere.

File: src/validation/error.messages.ts
Comment (jkayyala) [reply src/validation/error.messages.ts:137]: @mwestacott , it would contain OE and the number ex.OE123456 , but will rename the constant to INVALID_OE_NUMBER_LENGTH

File: src/routes/index.ts
Comment (corybudd-ch) [inline src/routes/index.ts:148]: @jkayyala could you check this [commit](https://github.com/companieshouse/overseas-entities-web/pull/482/commits/47b91012d18901a07c99550c56b05f83fb17cc2d) in UAR-91 branch where I separated out the routes/config for the update journey. As part of your rebase/merge I think you now have both declaration of routes and the validator needs to passed in through the route defs at the bottom of the file

File: src/validation/error.messages.ts
Comment (mwestacott) [reply src/validation/error.messages.ts:137]: 👍 thanks for clarifying.

File: src/middleware/authentication.middleware.ts
Comment (dgolubevs) [inline src/middleware/authentication.middleware.ts:15]: better 
```
const returnUrl =   req.path.startsWith(UPDATE_LANDING_URL)
   ? OVERSEAS_ENTITY_QUERY_URL
   : SOLD_LAND_FILTER_URL;
   
return res.redirect(`/signin?return_to=${returnUrl}`);
```

File: src/validation/overseas.entity.query.validation.ts
Comment (dgolubevs) [reply src/validation/overseas.entity.query.validation.ts:6]: it is extremely long line please sread by line
```

body("oe_number")
    .not().isEmpty({ ignore_whitespace: true })
    .withMessage(ErrorMessages.OE_QUERY_NAME)
    .isLength({ min: 8, max: 8 })
    .withMessage(ErrorMessages.INVALID_OE_LENGTH),
```

File: src/validation/error.messages.ts
Comment (dgolubevs) [reply src/validation/error.messages.ts:137]: Length can be only number number, it is not improve readabitlity

File: test/controllers/update/overseas.entity.query.controller.spec.ts
Comment (dgolubevs) [inline test/controllers/update/overseas.entity.query.controller.spec.ts:96]: should not be spaces ```{oe_number: ''} ```

File: test/middleware/authentication.middleware.spec.ts
Comment (dgolubevs) [inline test/middleware/authentication.middleware.spec.ts:12]: line too long
```

import {
   LANDING_URL,
   UPDATE_LANDING_URL,
   SOLD_LAND_FILTER_URL,
   OVERSEAS_ENTITY_QUERY_URL ,
} from '../../src/config';
```

File: test/controllers/update/overseas.entity.query.controller.spec.ts
Comment (jkayyala) [reply test/controllers/update/overseas.entity.query.controller.spec.ts:96]: @dgolubevs , lint:fix is configured to add spaces and can see the same format with spaces being used in other files also (ex. secure.register.filter.controller.spec.ts)

**Review by mwestacott — APPROVED** [review]
I'm happy with this so approving. Suggest getting approvals from others who commented before merging.

**Comment by jkayyala** [comment]
@dgolubevs , Can you please review, resolved most of the comments

**Review by dgolubevs — APPROVED** [review]

File: src/validation/overseas.entity.query.validation.ts
Comment (AndreKramer7654) [inline src/validation/overseas.entity.query.validation.ts:11]: The message should be invalid OE number! It's no longer a check for length only.

File: src/validation/regex/regex.validation.ts
Comment (AndreKramer7654) [inline src/validation/regex/regex.validation.ts:10]: VALID_OE_NUMBER_FORMAT

**Review by AndreKramer7654 — APPROVED** [review]

**Pull request merged**
